### ISSUE TITLE
Add drag-and-drop tab ordering

### ISSUE THREAD

**Issue by lisag** [AffiliateWP/affiliatewp-affiliate-area-tabs#8]
It would be cool to have a drag-and-drop option to move the order tabs around - we might also like to consider this for the default tabs?

**Comment by michaelbeil** [comment]
Great idea!

**Comment by michaelbeil** [comment]
Two requests for this recently.

**Comment by michaelbeil** [comment]
Another request https://secure.helpscout.net/conversation/240571815/0/

**Comment by michaelbeil** [comment]
Another request here: https://secure.helpscout.net/conversation/253120063/42768/?folderId=634609

**Comment by michaelbeil** [comment]
another: https://secure.helpscout.net/conversation/273845808/46182/?folderId=205524

**Comment by gspice** [comment]
Another request related to this, wants **Creatives** to be the default tab.  

https://secure.helpscout.net/conversation/303996926/50790/?folderId=1170842

**Comment by gspice** [comment]
Same as above, user getting wonky results when leaving only the Creatives tab active (to be the first, default tab) and adding the rest of the pages back using the shortcodes.  
https://secure.helpscout.net/conversation/422641577/65717/?folderId=1170842

**Comment by amdrew** [comment]
https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/47 added support for AffiliateWP v2.1.7's `affwp_get_affiliate_dashboard_tabs` filter. This means that tab priorities are now supported. A step in the right direction for completing this issue.

**Comment by amdrew** [comment]
@gspice In your comment above (https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/8#issuecomment-326046016) could the wonky results relate to this comment of mine? https://github.com/AffiliateWP/AffiliateWP/issues/1693#issuecomment-335065304

In summary, if you don't set the active tab via the filter (as per Drew's code example) then the first tab in the array (provided it is not disabled) will be shown as the "active" tab.

If so, this will be fixed once AffiliateWP v2.1.7 is released, ala: https://github.com/AffiliateWP/AffiliateWP/issues/1693. The tab with the lowest priority (first in the tab list) will be shown as the active tab.

**Comment by amdrew** [comment]
Here's a dedicated issue for setting the "default" or "active" tab:
https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/48

**Comment by amdrew** [comment]
@gspice Just realized you opened an issue for that quirk above (https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/46). I don't think after reading through that my comment is related.

**Comment by amdrew** [comment]
First pass: https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/pull/52

There's a few issues I'm aware of. Needs a bit more work and then it should be ready for a round of testing sometime tomorrow.

**Comment by amdrew** [comment]
New UI looks like this:

<img width="1231" alt="screenshot 2017-11-16 21 25 38" src="https://user-images.githubusercontent.com/52581/32881098-e31cea9c-cb14-11e7-8f3f-3ecb8679f362.png">

___

Adding a new custom tab:

<img width="1232" alt="screenshot 2017-11-16 21 25 49" src="https://user-images.githubusercontent.com/52581/32881096-e2ea4c40-cb14-11e7-8b94-e4d6062975c8.png">

<img width="1223" alt="screenshot 2017-11-16 21 26 11" src="https://user-images.githubusercontent.com/52581/32881095-e2bab156-cb14-11e7-8888-91bc519ba376.png">

___

Tabs can be reordered by clicking and dragging on the arrow handles. Shown is the custom tab now at a different position.

<img width="1225" alt="screenshot 2017-11-16 21 26 46" src="https://user-images.githubusercontent.com/52581/32881094-e289f30e-cb14-11e7-8889-6a1fbf003350.png">

___

**Comment by michaelbeil** [comment]
Looks great!

**Comment by pippinsplugins** [comment]
@amdrew Was getting rid of the open space to the left of the tabs not possible? Looks awesome otherwise!

**Comment by amdrew** [comment]
@pippinsplugins Have not yet tried, but on the list of things to look into 👍

**Comment by amdrew** [comment]
@AffiliateWP/core This is ready for testing.

As outlined on the PR, this should solve quite a few issues:

https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/49
https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/46
https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/41
https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/40
https://github.com/AffiliateWP/affiliatewp-affiliate-area-tabs/issues/8

## Notes

- Core tabs (added by AffiliateWP) cannot be deleted. They can however be hidden from showing in the Affiliate Area.

## Testing

With the new UI, drag and drop functionality, bug fixes etc, v1.2 is almost a complete rewrite. Listed below are some areas of focus I'd like to see tested:

#### Custom Tabs
- Reordering tabs
- Adding new custom tabs 
- Deleting tabs
- Hiding tabs

#### Core Tabs

- Reordering tabs
- Hiding tabs

#### Backwards compatibility

1.2 includes an upgrade script. It's important we test custom tabs with the current version (1.1.6) and check that everything works with the new version. All custom tabs should remain intact, and any current hidden tabs should also remain hidden.

With the current version (1.1.6), tabs can't be reordered. In 1.2, the upgrade routine will get any custom tabs and merge them with the default tabs from AffiliateWP, appending them to the end. Once upgraded, the customer can then re-order the tabs.

To test:

1. Checkout the `master` branch
2. With the current version (and old UI), add some custom tabs and hide some tabs.
3. Switch to the `issue/8` branch. The new UI should show the same existing custom tabs on the backend and front-end and also hide any tabs previously set.

Also check that there are no PHP notices outputted at any point during testing.

### Known issues

- Tabs added by add-ons (e.g. Direct Link Tracking) will show the `Tab Content` select menu in the admin.
- Adding a custom tab without a title still adds it
- The tab slug cannot be changed after initially adding the custom tag. In the future this will change each time the title of the tab is changed,

## Things left to do

- Set up grunt with CSS autoprefixer
- Minify scripts
- Clean up class names and CSS
- UI tweaks to match WP core
- Update readme to include new functionality
- See if we can remove the space to the left
- Uninstall script to clean up DB
- Make any JS text strings translatable
- Test custom tabs added via code

**Comment by amdrew** [comment]
Changed to be more consistent with WP core. Arrows appear "down" and change to "up" when clicked, like sidebars:

<img width="1020" alt="screenshot 2017-11-18 21 21 00" src="https://user-images.githubusercontent.com/52581/32978401-7d43ca86-cca6-11e7-84dc-ecf7749700a4.png">

<img width="1040" alt="screenshot 2017-11-18 21 21 18" src="https://user-images.githubusercontent.com/52581/32978408-a6405378-cca6-11e7-90a4-333b0a98b2f5.png">

**Comment by amdrew** [comment]
Tab counts are now moved to right hand side:

<img width="1027" alt="screenshot 2017-11-18 21 34 17" src="https://user-images.githubusercontent.com/52581/32978507-40f19746-cca8-11e7-9859-c9e817f5e734.png">

**Comment by amdrew** [comment]
Decided to make sure of the space to the left (for now) and added a way to quickly expand and collapse tabs:

<img width="1329" alt="screenshot 2017-11-19 22 35 05" src="https://user-images.githubusercontent.com/52581/32989399-fa263eca-cd79-11e7-8458-de332025e7a1.png">

This will be good for end-users and for testing.

**Comment by gspice** [comment]
So glad to have this feature on the way!

* First test, I used the master branch and hid a few of the default pages, and added one custom. I also have Direct Link Tracking, Order Details and Show Aff Coupons installed and active:

Admin view:
![aat-116-admin-settings](https://user-images.githubusercontent.com/3018245/33044652-8ec1e69a-cdfe-11e7-94de-f18d23f33517.png)

Affiliate view:
![aat-116-affiliate-view](https://user-images.githubusercontent.com/3018245/33044656-94d8e66e-cdfe-11e7-99a0-c5f2dc88c996.png)

* Next, I loaded the `Issue-8` branch and activated it.  Unfortunately it removed all of the standard pages, the custom page, and only is showing the add-on page mentioned above.  

Admin view:
![aat-1-2-admin-settings](https://user-images.githubusercontent.com/3018245/33044733-e443d43e-cdfe-11e7-8798-ef8432c00f16.png)

Affiliate view:
![aat-1-2-affiliate-view](https://user-images.githubusercontent.com/3018245/33044741-ed3cae94-cdfe-11e7-96ae-82d10218f4d2.png)

**Comment by amdrew** [comment]
@gspice If you refresh the admin page one more time, do the tabs come back? They should, if not, will look into it.

**Comment by gspice** [comment]
@amdrew I refreshed many times, including in Chrome dev tools the _Empty Cache and Hard Reload_ option.

**Comment by amdrew** [comment]
@gspice Ok thanks! Will look into it soon :)

**Comment by amdrew** [comment]
@gspice Ok, please pull down latest commits on `issue/8`.

To properly test the upgrade routine again you'll need to take a trip into your database and delete the following rows from your `wp_options` table:

`affwp_settings`
`affwp_aat_version_upgraded_from`
`affwp_aat_version`

Deleting the main settings row will wipe out any other affwp-related settings (which can easily be re-added) so bear that in mind. I have a dedicated sub-site locally for each add-on so nothing important is lost. Hit me up on slack if you're unsure on how to do the above and I'll guide you through it. 

Thanks!

**Comment by amdrew** [comment]
@pippinsplugins Going to leave the "space" to the left as-is. Few reasons:

1. I'm now using it for the expand/collapse link
2. The table is coming from WP core so can't see a way to remove it. On the plus side it's consistent with WP core setting pages.
3. It's consistent with our own settings, like the integrations tab for example (with space on left with just the word "integrations").

**Comment by gspice** [comment]
@amdrew Followed the steps, works great!  Testing a mix of hiding some defaults, moving them around and adding custom pages.

**Comment by pippinsplugins** [comment]
@amdrew I've just finished testing as well. I can't break it! Everything works really well.

**Comment by amdrew** [comment]
As discussed on Slack briefly I ran into a few issues with testing that will postpone this issue for the time-being.

1. Since `affwp_get_affiliate_area_tabs()` doesn't exist in AffiliateWP prior to version 2.1.7, there were fatal errors when upgrading AAT with AffiliateWP 2.1.6.1 and earlier. This has been since fixed.

The other issues first require a basic understanding of how AAT works:

AffiliateWP v2.1.7 introduced an array (with a filter) to hold the tabs outputted in the Affiliate Area, rather than each tab being hard-coded into the `dashboard.php` template file. AAT simply filters that array and replaces it with its own array of tabs, allowing the tabs to be shown in the order specified from the admin settings. Nice and simple.

2. Any tabs added via custom code won't instantly appear on the front-end, at least not until the settings have been saved. This is because the tabs shown on the front-end are pulled from the saved tabs in the DB, not from the array of tabs in AffiliateWP.

3. Add-ons such as Direct Linking, Show Affiliate Coupons and Order Details For Affiliates add custom tabs to the Affiliate Area. Though they can appear in the specified order on the front-end (after saving), if an admin decides to turn off DLT, or only for a specific affiliate, the tab(s) will still show on the front-end which is a bit problematic. They need to removed instantly.

`unset()` would probably solve the above two issues. We could look at the `affwp_get_affiliate_area_tabs()` array, compare that to the tabs array stored by AAT, and unset any tabs if they exist in AAT’s array, but not in `affwp_get_affiliate_area_tabs()`. We'd skip over any custom tabs added via the admin interface since these will have page IDs assigned. Custom tabs added via code, or added from an admin do not have a page ID assigned.

**Comment by amdrew** [comment]
Fixed the issues listed above and merged into master.

Some final notes:

## Custom tabs added via a code snippet

* The tab is appended to the very end of the current set of tabs, on both the frontend Affiliate Area and the AAT settings.
* The tab can be re-ordered just like any other tab and saved in its new position, which will be reflected on the Affiliate Area
* When the code snippet is removed, the custom tab is immediately removed from both the Affiliate Area, and also from AAT settings. Technically, the tab still exists in the tab array within the database settings, its only removed via unset(). Once the admin saves the tab settings however, its then removed for good. The only important aspect is that it does not show.
* The custom tab can be hidden using the “Hide tab in Affiliate Area” checkbox setting.
* The custom tab can be given a notice like the other add-ons by registering itself with AAT using the `affwp_aat_add_on_tabs` filter and passing a notice.

## Custom tabs added by add-ons
Example: Direct Link Tracking, Show Affiliate Coupons, Order Details For Affiliates

* Tabs added by add-ons now immediately disappear for an affiliate if they no longer have access, or if the add-on is completely disabled
* Tabs will always show in the AAT settings unless the add-on is disabled. Admins can re-order the tabs as usual even though some affiliates may not see the tab
* The tab notices have been updated to reflect the above. For example, DLT’s notice states "This tab has been added from the Direct Link Tracking add-on. Only affiliates with Direct Link Tracking enabled will see this tab in the Affiliate Area."
